### PR TITLE
Refactor `Distance` constructor, update docstring and error messages

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

This pull request contains two commits. The first refactors the `Distance` constructor by removing a level of indentation and reducing code duplication. The second commit implements (mostly) small changes to the error messages.

EDIT: I have added a third commit that updates the `Distance` docstring.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Review by mhvk — APPROVED** [review]
Thanks! Looks all OK to me, so approving, and can merge once tests pass.

**Comment by pllim** [comment]
Might need to fix the regex:

```
        with pytest.raises(ValueError, match='none of `value`, `z`, `distmod` or'):
>           Distance(cosmology=WMAP5, unit=u.km)
E           AssertionError: Regex pattern 'none of `value`, `z`, `distmod` or' does not match 'none of `value`, `z`, `distmod`, or `parallax` were given to Distance constructor'.
```

**Comment by eerovaher** [comment]
While working on the `Distance` constructor I noticed that the `unit` keyword has no effect if the `Distance` is being constructed from a parallax. Fixing that bug should be very simple, but the lines that would have to be edited are also being modified by this pull request. It just occurred to me that (unless I'm mistaken) backporting the upcoming bugfix would be easier if this pull request is not merged first, so I'll delay fixing the typo I made in one of the regex patterns for now.

File: astropy/coordinates/distances.py
Comment (eerovaher) [inline astropy/coordinates/distances.py:70]: The `Distance` constructor no longer raises a `UnitsError` itself, but trying to use an invalid unit does lead to a `UnitsError` being raised elsewhere.

File: astropy/coordinates/distances.py
Comment (eerovaher) [inline astropy/coordinates/distances.py:93]: Apparently it is possible to specify a distance modulus without a unit. It might be better to be stricter about enforcing a unit here, but even if it is then making the change is beyond the scope of this pull request.

File: astropy/coordinates/distances.py
Comment (mhvk) [reply astropy/coordinates/distances.py:70]: I think it is OK to leave it here, though.

File: astropy/coordinates/distances.py
Comment (mhvk) [reply astropy/coordinates/distances.py:93]: I see that it just assumes a default unit of `u.mag`. I think that is probably not worth changing. Though my sense would be to use `24.47*u.mag` in the example, as it is more explicit.

File: astropy/coordinates/distances.py
Comment (mhvk) [inline astropy/coordinates/distances.py:95]: Shall we delete this one? It should be obvious that that works!

**Review by mhvk — APPROVED** [review]
This looks completely good as is, but I have two very minimal comments that you may want to think about. Either way, (re-)approved!

File: astropy/coordinates/distances.py
Comment (eerovaher) [reply astropy/coordinates/distances.py]: I have added the unit to the example.

File: astropy/coordinates/distances.py
Comment (eerovaher) [reply astropy/coordinates/distances.py]: I have removed this example.

**Comment by mhvk** [comment]
Failures unrelated (though numpy-dev seems bad...), so merging...

**Comment by pllim** [comment]
Not sure what happened but after the merge of this PR, CI on `main` broke.

`ERROR: InvocationError for command .../pytest --pyargs astropy .../astropy/docs --durations=50 (exited with code -6 (SIGABRT)) `

Should we revert?

**Comment by pllim** [comment]
https://github.com/astropy/astropy/actions/runs/1560183565

**Comment by mhvk** [comment]
That is bizarre, especially as it breaks while doing unit tests, even though the `astropy.units` module has not a single test with `Distance` in it...  But good idea just to see whether without this PR tests do pass...

Let me quickly push something else I was planning to do anyway, so we can check if master is truly broken...

**Comment by pllim** [comment]
pytest-doctestplus also just got a new release... @saimn , any chance the new `.md` support would break astropy doctest?

**Comment by mhvk** [comment]
See #12579 for a non-reversion test...

**Comment by mhvk** [comment]
Now if only I had been smart enough to rebase on master before sending it in...  Now done, so yet more time to wait whether tests pass...

**Comment by saimn** [comment]
@pllim - It's failing on `test_quantity_ufuncs.py` so should not be related to pytest-doctestplus I guess.

**Comment by mhvk** [comment]
@pllim - #12579 passed the same test, so it looks like it was some kind of weird fluke

**Comment by pllim** [comment]
OK, let me restart the `main` CI. Just in case.

**Comment by pllim** [comment]
Yeah, okay, must have been a fluke. Sorry for the noise!

**Pull request merged**
